src/cpp/session/modules/SessionFind.cpp

```diff
@@ -629,87 +630,108 @@ class GrepOperation : public boost::enable_shared_from_this<GrepOperation>
-                                     pReplaceMatchOff, &lineSuccess);
+                     addReplaceErrorMessage(error.asString(), pErrorMessage, pReplaceMatchOn,
+                        pReplaceMatchOff, &lineSuccess);
                      return error;
                   }
                }
-               if (findResults().preview())
-                  replaceString.insert(0, pLineInfo->decodedContents.substr(matchOn, matchSize));
+               else // perform replace
+               {
+                  pProgress->addUnits(1);
+                  if (findResults().replaceRegex())
+                     error = replacer.replaceRegexWithRegex(matchOn, matchOff, searchPattern,
+                        replacePattern, &newLine, &replaceMatchOff);
+                  else if (findResults().regex())
+                     error = replacer.replaceRegexWithLiteral(matchOn, matchOff, searchPattern,
+                        replacePattern, &newLine, &replaceMatchOff);
+                  else
+                     replacer.replaceLiteralWithLiteral(matchOn, matchOff, replacePattern, &newLine,
+                        &replaceMatchOff);
 
-               Replacer replacer(findResults().ignoreCase());
-               if (findResults().regex() &&
-                   !findResults().replaceRegex())
-                  error = replacer.replaceRegexWithLiteral(matchOn, matchOff, &newLine, &searchPattern,
-                     &replacePattern, &replaceMatchOff);
-               else
-                  replacer.replaceLiteralWithLiteral(matchOn, matchOff, &newLine, &replaceString,
-                     &replaceMatchOff);
-               if (error)
-                  addErrorMessage(error.asString(), pErrorMessage,
-                     pReplaceMatchOn, pReplaceMatchOff, &lineSuccess);
+                  if (error)
+                     addReplaceErrorMessage(error.asString(), pErrorMessage, pReplaceMatchOn,
+                        pReplaceMatchOff, &lineSuccess);
+               }
+
+               // Handle side-effects when replace is successful
                if (lineSuccess)
                {
                   pLineInfo->decodedContents = newLine;
 
                   // if multiple replaces in line, readjust previous match numbers
+                  size_t replaceSize = replaceMatchOff - matchOn;
                   if (pReplaceMatchOn->getSize() > 0 &&
-                      matchSize != replaceString.size())
+                      matchSize != replaceSize)
                   {
                      json::Array tempMatchOn(*pReplaceMatchOn);
                      json::Array tempMatchOff(*pReplaceMatchOff);
                      pReplaceMatchOn->clear();
                      pReplaceMatchOff->clear();
 
-                     // put back in reverse order
-                     int offset(gsl::narrow_cast<int>(replaceString.size() - matchSize));
+                     // put back in reverse order for the frontend
```

Comment:
To make sure I understand correctly - the backend and the frontend want the results in different orders because the backend wants to iterate through them from last result in file to first result in file and the frontend wants to display them from first result in file to last? If that's correct, could you add a comment somewhere in the code to that effect (or whatever the real reason is)? Sorry if it's already there and I've missed it

Reply:
That's correct, except it's first/last match per line, not per file. We always read the lines in the order they appear in the file. I updated the comment to be more specific.